libs/model/src/services/openai/generateTokenIdea.ts

```diff
@@ -0,0 +1,175 @@
+import { OpenAI } from 'openai';
+import {
+  ChatCompletionMessage,
+  ChatCompletionUserMessageParam,
+} from 'openai/resources/index.mjs';
+import { v4 as uuidv4 } from 'uuid';
+import { config } from '../../config';
+import { models } from '../../database';
+import { TokenInstance } from '../../models/token';
+
+type TokenIdea = {
+  name: string;
+  symbol: string;
+  description: string;
+  imageURL: string;
+};
+
+type GenerateTokenIdeaProps = {
+  ideaPrompt?: string;
+};
+
+type GenerateTokenIdeaResponse = {
+  tokenIdea?: TokenIdea;
+  error?: string;
+};
+
+const TOKEN_AI_PROMPTS_CONFIG = {
+  name: (ideaPrompt?: string) => `
+      Please ${ideaPrompt ? `use this prompt in quotation "${ideaPrompt}" to ` : ''}
+      create a random idea for a memecoin cryptocurrency, have the token name informed by current events, 
+      and historical internet based humor. Make it hyper opinionated and subversive or topical. Provide just the 
+      token name without any acronym or symbol. Make it either cringe or amazing. Restrict your answer to between 
+      6 and 25 characters.
+    `,
+  symbol: `
+      Create a symbol for this token (symbol) of 3-6 characters. The symbol should not include special characters 
+      or emojis. Restrict your answer to between 3 and 6 characters.`,
+  description: `
+      Provide a description for the token. Make it funny and keep it in the vein of ironic, sardonic, twitter personae.
+      There are no emoji restrictions. Keep the description to 3 sentences. DO NOT BE OVERLY POSITIVE about global 
+      phenomenon, only the asset itself. Restrict your answer to between 1 and 3 sentences and less than 180 characters.
+    `,
+  image: (name: string, symbol: string) => `
+      Please create an image for a web3 token named "${name}", having a symbol of "${symbol}". 
+    `,
+};
+
+const TokenErrors = {
+  OpenAINotConfigured: 'OpenAI key not configured',
+  OpenAIInitFailed: 'OpenAI initialization failed',
+  RequestFailed: 'failed to generate complete token idea generation',
+  IdeaPromptVoilatesSecurityPolicy:
+    'provided `ideaPrompt` voilates content security policy',
+  ImageGenerationFailure: 'failed to generate image for token idea',
+};
+
+// we have to maintain this to have "conversational" context with OpenAI
+const convoHistory: (ChatCompletionMessage | ChatCompletionUserMessageParam)[] =
+  [];
+
+const chatWithOpenAI = async (prompt = '', openai: OpenAI) => {
+  convoHistory.push({ role: 'user', content: prompt }); // user msg
+
+  const response = await openai.chat.completions.create({
+    model: 'gpt-3.5-turbo',
+    messages: convoHistory,
+  });
+  convoHistory.push(response.choices[0].message); // assistant msg
+
+  return (response.choices[0].message.content || 'NO_RESPONSE').replace(
+    /^"|"$/g, // sometimes openAI adds `"` at the start/end of the response + tweaking the prompt also doesn't help
+    '',
+  );
+};
+
+const generateTokenIdea = async function* ({
```

Comment:
whats the rationale to use generator here?

Reply:
TLDR: To stream data as its generated. 

Long answer:
- we make 4 different requests to chatgpt and 1 request to upload generated image to s3 = 5 requests in total
- each request depends on the responses from previous request
- we don't want the user to wait a full 15-20 sec and then show data on their screens so we stream the chunks of data as we generate them

Reply:
thanks for explanation, makes sense